### PR TITLE
Add GPT-4 Vision API wrapper

### PR DESCRIPTION
**Pull request by jmanhype**
Introduce a new GPT4Vision class in visionopenai.py that wraps the GPT-4 Vision API. This abstraction layer simplifies the process of making requests to the API for analyzing images. Key functionality includes:

Encoding images to base64
Preparing image metadata in the required format
Calculating token costs for images
Making API requests with both text and image prompts Error handling and validation

### CODE REVIEW

**Comment by isaacbmiller** [comment]
Also is there any way to add tests for this?

**Review by isaacbmiller — COMMENTED** [review]
Also run "ruff check . --fix-only"

**Comment by isaacbmiller** [comment]
> I have a unittest created.

Where?

**Comment by landworth** [comment]
Adding GPT-4 Vision capabilties would be super useful! Any progress with fixing the failing check here? 🙂

**Comment by chitianhao** [comment]
Did you try whether this really works or not?

What does the Signature look like to pass an image?

**Comment by arnavsinghvi11** [comment]
Thanks @jmanhype for the contribution! Could you resolve https://github.com/stanfordnlp/dspy/pull/682#pullrequestreview-1950064068? 

Also, it would be great if you add documentation for this under the [documentation folder as with the other LMs](https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients). Thanks!

**Comment by jmanhype** [comment]
I just made those updates

File: dsp/modules/visionopenai.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+import logging
+import requests
+import base64
+from typing import Any, List, Optional, Union
+
+import openai
```
[the remark this answers is not in the document]
Comment (arnavsinghvi11) [reply dsp/modules/visionopenai.py:6]: @jmanhype tagging you here for this change

**Comment by arnavsinghvi11** [comment]
@jmanhype 

> Also run "ruff check . --fix-only"



>  add documentation for this under the [documentation folder as with the other LMs](https://github.com/stanfordnlp/dspy/tree/main/docs/api/language_model_clients). Thanks!

**Comment by jmanhype** [comment]
Done

**Comment by arnavsinghvi11** [comment]
@jmanhype did you push? don't see new commits.

**Comment by jmanhype** [comment]
ok everything has been updated

File: dsp/modules/visionopenai.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        image_data = []
+        for image in images:
+            if is_url:
+                if "url" not in image:
+                    raise ValueError("Image dictionary must contain a 'url' key for URL-based images")
+                url = image["url"]
+                if not url.startswith(("http://", "https://")):
+                    raise ValueError("Invalid image URL")
+                # You can perform additional URL validation if needed
```
Comment (arnavsinghvi11) [inline dsp/modules/visionopenai.py:61]: remove extra comments

File: dsp/modules/visionopenai.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            else:
+                if "data" not in image:
+                    raise ValueError("Image dictionary must contain a 'data' key for base64-encoded images")
+                base64_data = image["data"]
+                if not base64_data.startswith("data:image/"):
+                    raise ValueError("Invalid base64-encoded image data")
+                # You can perform additional base64 validation if needed
```
Comment (arnavsinghvi11) [inline dsp/modules/visionopenai.py:68]: remove extra comments

File: dsp/modules/visionopenai.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        Args:
+            prompts (List[dict]): A list of prompts with text or images.
+            **kwargs: Additional arguments for the API call.
+
+        Returns:
+            List[Any]: The list of responses from the model.
+        """
+        return self.request(prompts, **kwargs)
+
+    # Add any additional methods needed for specific GPT-4 Vision API functionalities.
```
Comment (arnavsinghvi11) [inline dsp/modules/visionopenai.py:177]: remove extra comments (or leave as TODO)

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Hi @jmanhype still needs `ruff check . --fix-only` to be run.

Also, to properly call `dspy.GPT4Vision`, you need to import it as the other LM modules:
- [dsp module initialization](https://github.com/stanfordnlp/dspy/blob/e655d2233fcd70bbb3efea474086954b420f9b3a/dsp/modules/__init__.py#L11)
- [dspy module initialization](https://github.com/stanfordnlp/dspy/blob/e655d2233fcd70bbb3efea474086954b420f9b3a/dspy/__init__.py#L36)

**Comment by jmanhype** [comment]
Ruff check is false positive it passes locally

**Comment by AyushPibit1234** [comment]
@arnavsinghvi11 Is there any update on this? 
cc : @jmanhype

**Comment by jmanhype** [comment]
Just needs to be merged the ruff error is a false positive

**Comment by isaacbmiller** [comment]
Hey @jmanhype I did actually find a ruff error and I added GPT4Vision to the __init__ files in dsp and dspy like @arnavsinghvi11 mentioned.

See this branch: 
[https://github.com/stanfordnlp/dspy/tree/jmanhype/main](https://github.com/stanfordnlp/dspy/tree/jmanhype/main)

**Pull request closed without merging**